File: 4. Longest Palindromic Substring/sol.cpp
#include <iostream>
#include <string>
using namespace std;

// Function to expand around center and find the longest palindrome
string expandAroundCenter(string s, int left, int right) {
    while (left >= 0 && right < s.length() && s[left] == s[right]) {
        left--;
        right++;
    }
    return s.substr(left + 1, right - left - 1);
}

// Function to find the longest palindromic substring
string longestPalindrome(string s) {
    if (s.empty()) return "";
    
    string longest = "";
    
    for (int i = 0; i < s.length(); i++) {
        string oddPal = expandAroundCenter(s, i, i);     // Odd length palindrome
        string evenPal = expandAroundCenter(s, i, i + 1); // Even length palindrome
        
        if (oddPal.length() > longest.length()) longest = oddPal;
        if (evenPal.length() > longest.length()) longest = evenPal;
    }
    
    return longest;
}

// Main function
int main() {
    string str = "babad";
    cout << "Longest Palindromic Substring: " << longestPalindrome(str) << endl;
    return 0;
}

File: 4. Longest Palindromic Substring/sol.java
class sopl {
    
    // Function to expand around center and find the longest palindrome
    public static String expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return s.substring(left + 1, right);
    }

    // Function to find the longest palindromic substring
    public static String longestPalindrome(String s) {
        if (s == null || s.length() == 0) return "";
        
        String longest = "";
        
        for (int i = 0; i < s.length(); i++) {
            String oddPal = expandAroundCenter(s, i, i);      // Odd length palindrome
            String evenPal = expandAroundCenter(s, i, i + 1); // Even length palindrome
            
            if (oddPal.length() > longest.length()) longest = oddPal;
            if (evenPal.length() > longest.length()) longest = evenPal;
        }
        
        return longest;
    }

    // Main function
    public static void main(String[] args) {
        String str = "babad";
        System.out.println("Longest Palindromic Substring: " + longestPalindrome(str));
    }
}
